src/bun.js/node/types.zig

```diff
@@ -1835,548 +1849,2876 @@ pub const Emitter = struct {
-    const StringBuilder = @import("../../string_builder.zig");
-    pub const code = @embedFile("../path.exports.js");
+    pub const delimiter = CHAR_COLON;
+    pub const sep_posix = CHAR_FORWARD_SLASH;
+    pub const sep_windows = CHAR_BACKWARD_SLASH;
+    pub const sep_str_posix = CHAR_STR_FORWARD_SLASH;
+    pub const sep_str_windows = CHAR_STR_BACKWARD_SLASH;
+    
+    inline fn ParsedPathImpl(comptime T: type, root: []const T, dir: []const T, base: []const T, ext: []const T, _name: []const T) ParsedPathUnion {
+        const ParsedPathT = if (@TypeOf(T) != u16) ParsedPath else ParsedPathW;
+        const result = ParsedPathT{ .root = root, .dir = dir, .base = base, .ext = ext, .name = _name };
+        return if (@TypeOf(T) != u16) ParsedPathUnion{ .u8 = result } else ParsedPathUnion{ .u16 = result };
+    }
 
-    pub fn create(globalObject: *JSC.JSGlobalObject, isWindows: bool) callconv(.C) JSC.JSValue {
-        return shim.cppFn("create", .{ globalObject, isWindows });
+    /// Taken from Zig 0.11.0 zig/src/resinator/rc.zig
+    /// https://github.com/ziglang/zig/blob/776cd673f206099012d789fd5d05d49dd72b9faa/src/resinator/rc.zig#L266
+    ///
+    /// Compares ASCII values case-insensitively, non-ASCII values are compared directly
+    fn eqlIgnoreCaseImpl(comptime T: type, a: []const T, b: []const T) bool {
+        if (@TypeOf(T) != u16) {
+            return std.ascii.eqlIgnoreCase(a, b);
+        }
+        if (a.len != b.len) return false;
+        for (a, b) |a_c, b_c| {
+            if (a_c < 128) {
+                if (std.ascii.toLower(@intCast(a_c)) != std.ascii.toLower(@intCast(b_c))) return false;
+            } else {
+                if (a_c != b_c) return false;
+            }
+        }
+        return true;
     }
 
-    pub fn basename(globalThis: *JSC.JSGlobalObject, isWindows: bool, args_ptr: [*]JSC.JSValue, args_len: u16) callconv(.C) JSC.JSValue {
-        if (comptime is_bindgen) return JSC.JSValue.jsUndefined();
-        if (args_len == 0) {
-            return JSC.toInvalidArguments("path is required", .{}, globalThis);
+    /// Based on Node v21.6.1 private helper formatExt:
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L130C10-L130C19
+    inline fn formatExtImpl(comptime T: type, ext: []const T, buf: []T) []const T {
+        const len = ext.len;
+        if (len == 0) {
+            return "";
         }
-        var stack_fallback = std.heap.stackFallback(4096, JSC.getAllocator(globalThis));
-        const allocator = stack_fallback.get();
+        if (ext[0] == CHAR_DOT) {
+            return ext;
+        }
+        const bufSize = len + 1;
+        buf[0] = CHAR_DOT;
+        @memcpy(buf[1 ..bufSize], ext);
+        return buf[0 ..bufSize];
+    }
 
-        var arguments: []JSC.JSValue = args_ptr[0..args_len];
-        var path = arguments[0].toSlice(globalThis, allocator);
+    inline fn getCwd(buf: []u8) []const u8 {
```

Comment:
`bun.getcwd` is `std.os.getcwd`:
https://github.com/oven-sh/bun/blob/5934b17f0013410d3b7111ca3f98d7a79da9698b/src/bun.zig#L1118.

I opted for `std.c.getcwd` because that's what Process uses:
https://github.com/oven-sh/bun/blob/5934b17f0013410d3b7111ca3f98d7a79da9698b/src/bun.js/node/types.zig#L2552

Which calls `Syscall.getcwd`, which calls `std.c.getcwd`:
https://github.com/oven-sh/bun/blob/5934b17f0013410d3b7111ca3f98d7a79da9698b/src/sys.zig#L153-L161

By using the same API I can expect the same result. We should pick one variant.

Reply:
this is revealing existing bad code

- for windows, std.os.getcwd should be used so it uses GetCurrentDirectory instead of ucrt
- on posix, `os.getcwd` calls `c.getcwd`
- the null byte seems to have no purpose as it is overwritten instantly (buffer size indicated by the second param, see https://man7.org/linux/man-pages/man3/getcwd.3.html and https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getcurrentdirectory)

i still believe we should remove this function

Reply:
> the null byte seems to have no purpose as it is overwritten instantly (buffer size indicated by the second param, 

In C-land which is what `Syscall.getcwd` is calling into:
https://github.com/oven-sh/bun/blob/48ed1dfb65b77f29a4969a340f949f69ecfb9a64/src/sys.zig#L153-L156

My understanding is the null terminator is a way to indicate the end of a string. So by setting it to `0` it's essentially saying its a fresh start.

> i still believe we should remove this function

I'll change both to using the bun.getcwd form.